Fatal Python error: GIL held when using fsspec with ArrowFSWrapper

### What happens?

I want to register and use a ArrowFSWrapper-based fsspec filesystem (https://arrow.apache.org/docs/python/filesystems.html#using-arrow-filesystems-with-fsspec) with DuckDB, using the new fsspec feature in the Python client (https://github.com/duckdb/duckdb/pull/5829).

I can register the filesystem successfully, using `duckdb_cursor.register_filesystem(local_fsspec)`, but when I try to read data from that filesystem, the Python process runs into a fatal GIL error. 

### To Reproduce

This is a self-contained example to reproduce the issue.

```
import duckdb
from pyarrow import fs
from fsspec.implementations.arrow import ArrowFSWrapper
local = fs.LocalFileSystem()
local_fsspec = ArrowFSWrapper(local)
local_fsspec.protocol = "local"
with local_fsspec.open("test.csv", mode='w') as f:
    f.write("a,b,c\n")
    f.write("1,2,3\n")
    f.write("4,5,6\n")
duckdb_cursor = duckdb.connect()
duckdb_cursor.register_filesystem(local_fsspec) 
duckdb_cursor.execute(f"select * from read_csv_auto('local://test.csv')")

-->
Fatal Python error: PyThreadState_Get: the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)
Python runtime state: initialized

Current thread 0x00007f137f309200 (most recent call first):
  File "<stdin>", line 1 in <module>
```

duckdb version: '0.6.2-dev1759'
fsspec version: '2022.11.0'
pyarrow version: '11.0.0'

### OS:

Windows

### DuckDB Version:

0.6.2-dev1759

### DuckDB Client:

Python

### Full Name:

Till Döhmen

### Affiliation:

Hopsworks AB

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

Okay, as discussed on Discord, this was indeed a duplicate of the GC issue. However, I can into the below issue:
```py
from pyarrow import fs
from fsspec.implementations.arrow import ArrowFSWrapper

local = fs.LocalFileSystem()
local_fsspec = ArrowFSWrapper(local, skip_instance_cache=True)
local_fsspec.protocol = "local"
with local_fsspec.open("test.csv", mode='w') as f:
    f.write("a,b,c\n")
    f.write("1,2,3\n")
    f.write("4,5,6\n")
with local_fsspec.open("test.csv", seekable=True) as f:
    assert f.stream.seekable(), 'stream not seekable'
    assert f.seekable(), 'file not seekable'
```
which gives
```
Traceback (most recent call last):
  File "/home/me/duckdb/tools/pythonpkg/thing.py", line 13, in <module>
    assert f.seekable(), 'file not seekable'
AssertionError: file not seekable
```
which seems very odd, as the seekable method on ArrowFile should just be a straight through call?

--- a/tools/pythonpkg/src/include/duckdb_python/pyfilesystem.hpp
+++ b/tools/pythonpkg/src/include/duckdb_python/pyfilesystem.hpp
@@ -22,6 +22,10 @@ public:
 class PythonFileHandle : public FileHandle {
 public:
 	PythonFileHandle(FileSystem &file_system, const string &path, const py::object handle);
+	~PythonFileHandle() {
+		PythonGILWrapper gil;
+		handle.dec_ref();
+	}
 
 	void Close() override {
 		PythonGILWrapper gil;

--- a/tools/pythonpkg/src/include/duckdb_python/pyfilesystem.hpp
+++ b/tools/pythonpkg/src/include/duckdb_python/pyfilesystem.hpp
@@ -22,11 +22,7 @@ public:
 class PythonFileHandle : public FileHandle {
 public:
 	PythonFileHandle(FileSystem &file_system, const string &path, const py::object handle);
-	~PythonFileHandle() {
-		PythonGILWrapper gil;
-		handle.dec_ref();
-	}
-
+	~PythonFileHandle() override;
 	void Close() override {
 		PythonGILWrapper gil;
 		handle.attr("close")();

--- a/tools/pythonpkg/src/pyfilesystem.cpp
+++ b/tools/pythonpkg/src/pyfilesystem.cpp
@@ -31,7 +31,8 @@ unique_ptr<FileHandle> PythonFilesystem::OpenFile(const string &path, uint8_t fl
 		throw InvalidInputException("%s: unsupported file flags", GetName());
 	}
 
-	const auto &handle = filesystem.attr("open")(py::str(stripPrefix(path)), py::str(flags_s));
+	// `seekable` is passed here for `ArrowFSWrapper`, other implementations seem happy enough to ignore it
+	const auto &handle = filesystem.attr("open")(py::str(stripPrefix(path)), py::str(flags_s), py::arg("seekable") = true);
 	return make_unique<PythonFileHandle>(*this, path, handle);
 }
 

--- a/tools/pythonpkg/src/pyfilesystem.cpp
+++ b/tools/pythonpkg/src/pyfilesystem.cpp
@@ -32,7 +32,8 @@ unique_ptr<FileHandle> PythonFilesystem::OpenFile(const string &path, uint8_t fl
 	}
 
 	// `seekable` is passed here for `ArrowFSWrapper`, other implementations seem happy enough to ignore it
-	const auto &handle = filesystem.attr("open")(py::str(stripPrefix(path)), py::str(flags_s), py::arg("seekable") = true);
+	const auto &handle =
+	    filesystem.attr("open")(py::str(stripPrefix(path)), py::str(flags_s), py::arg("seekable") = true);
 	return make_unique<PythonFileHandle>(*this, path, handle);
 }
 

--- a/tools/pythonpkg/src/pyfilesystem.cpp
+++ b/tools/pythonpkg/src/pyfilesystem.cpp
@@ -9,6 +9,11 @@ namespace duckdb {
 PythonFileHandle::PythonFileHandle(FileSystem &file_system, const string &path, const py::object handle)
     : FileHandle(file_system, path), handle(handle) {
 }
+PythonFileHandle::~PythonFileHandle() {
+	PythonGILWrapper gil;
+	handle.dec_ref();
+	handle.release();
+}
 
 unique_ptr<FileHandle> PythonFilesystem::OpenFile(const string &path, uint8_t flags, FileLockType lock,
                                                   FileCompressionType compression, FileOpener *opener) {

--- a/tools/pythonpkg/tests/fast/test_filesystem.py
+++ b/tools/pythonpkg/tests/fast/test_filesystem.py
@@ -4,6 +4,7 @@ from pathlib import Path
 from shutil import copyfileobj
 from typing import Callable
 
+import duckdb
 from duckdb import DuckDBPyConnection, InvalidInputException
 from pytest import raises, importorskip, fixture, MonkeyPatch
 
@@ -104,3 +105,22 @@ class TestPythonFilesystem:
 
         with raises(ModuleNotFoundError):
             duckdb_cursor.register_filesystem(None)
+
+    def test_arrow_fs_wrapper(self):
+        fs = importorskip('pyarrow.fs')
+        from fsspec.implementations.arrow import ArrowFSWrapper
+
+        local = fs.LocalFileSystem()
+        local_fsspec = ArrowFSWrapper(local, skip_instance_cache=True)
+        local_fsspec.protocol = "local"
+        filename = "test.csv"
+        with local_fsspec.open(filename, mode='w') as f:
+            f.write("a,b,c\n")
+            f.write("1,2,3\n")
+            f.write("4,5,6\n")
+
+        duckdb_cursor = duckdb.connect()
+        duckdb_cursor.register_filesystem(local_fsspec)
+        duckdb_cursor.execute(f"select * from read_csv_auto('local://{filename}', header=true)")
+
+        assert duckdb_cursor.fetchall() == [(1, 2, 3), (4, 5, 6)]

--- a/tools/pythonpkg/tests/fast/test_filesystem.py
+++ b/tools/pythonpkg/tests/fast/test_filesystem.py
@@ -106,14 +106,14 @@ class TestPythonFilesystem:
         with raises(ModuleNotFoundError):
             duckdb_cursor.register_filesystem(None)
 
-    def test_arrow_fs_wrapper(self):
+    def test_arrow_fs_wrapper(self, tmp_path: Path):
         fs = importorskip('pyarrow.fs')
         from fsspec.implementations.arrow import ArrowFSWrapper
 
         local = fs.LocalFileSystem()
         local_fsspec = ArrowFSWrapper(local, skip_instance_cache=True)
         local_fsspec.protocol = "local"
-        filename = "test.csv"
+        filename = str(tmp_path / "test.csv")
         with local_fsspec.open(filename, mode='w') as f:
             f.write("a,b,c\n")
             f.write("1,2,3\n")

--- a/tools/pythonpkg/tests/fast/test_filesystem.py
+++ b/tools/pythonpkg/tests/fast/test_filesystem.py
@@ -6,7 +6,7 @@ from typing import Callable
 
 import duckdb
 from duckdb import DuckDBPyConnection, InvalidInputException
-from pytest import raises, importorskip, fixture, MonkeyPatch
+from pytest import raises, importorskip, fixture, MonkeyPatch, mark
 
 importorskip('fsspec', '2022.11.0')
 from fsspec import filesystem, AbstractFileSystem
@@ -106,6 +106,7 @@ class TestPythonFilesystem:
         with raises(ModuleNotFoundError):
             duckdb_cursor.register_filesystem(None)
 
+    @mark.skipif(sys.version_info < (3, 8), reason="ArrowFSWrapper requires python 3.8 or higher")
     def test_arrow_fs_wrapper(self, tmp_path: Path):
         fs = importorskip('pyarrow.fs')
         from fsspec.implementations.arrow import ArrowFSWrapper

